File: sdks/typescript/mcp-core/src/config-log.ts
/**
 * One-line MCP config summary to stderr, once per process.
 *
 * The message body comes from the Rust `mcpConfigLog` op. The
 * once-per-process guard stays host-side.
 */

import { callMcpSyncOp } from './native-mcp'

let logged = false

export type McpConfigLogInput = {
  apiBaseUrl: string
  productRef: string
  publicBaseUrl: string
}

export function mcpConfigLogMessage(config: McpConfigLogInput): string {
  return callMcpSyncOp<{ message: string }>('mcpConfigLog', {
    apiBaseUrl: config.apiBaseUrl,
    productRef: config.productRef,
    publicBaseUrl: config.publicBaseUrl,
  }).message
}

/**
 * Emit `[solvapay] mcp config …` once. Safe to call from every construction
 * path; subsequent calls are no-ops.
 */
export function logMcpConfigOnce(config: McpConfigLogInput): void {
  if (logged) return
  logged = true
  console.warn(mcpConfigLogMessage(config))
}

/** Test-only: reset the once-per-process guard. */
export function resetMcpConfigLogForTests(): void {
  logged = false
}

File: sdks/python-mcp/python/solvapay_mcp/oauth/config_log.py
from __future__ import annotations

import logging

_LOGGER = logging.getLogger("solvapay")
_logged = False


def mcp_config_log_message(api_base_url: str, product_ref: str, public_base_url: str) -> str:
    from solvapay_mcp.core import call

    value = call(
        "mcpConfigLog",
        {
            "apiBaseUrl": api_base_url,
            "productRef": product_ref,
            "publicBaseUrl": public_base_url,
        },
    )
    if not isinstance(value, dict) or not isinstance(value.get("message"), str):
        raise TypeError("mcpConfigLog did not return a message")
    return str(value["message"])


def log_mcp_config_once(api_base_url: str, product_ref: str, public_base_url: str) -> None:
    global _logged
    if _logged:
        return
    _logged = True
    _LOGGER.warning(mcp_config_log_message(api_base_url, product_ref, public_base_url))


def reset_mcp_config_log_for_tests() -> None:
    global _logged
    _logged = False
